UI - onboarding styling and colors

danielnordh opened the pull request:
### Description

This PR makes changes to the onboarding screen styling and colors.

<img width="982" alt="image" src="https://github.com/user-attachments/assets/685b61ae-fd9c-4d5c-ac87-31919a7b0d49">

Before / After light mode / After dark mode

### Changelog notice

- Updated onboarding screen UI with styling and color
- Moved network settings to separate screen
- Moved wallet import to separate screen

### Checklists

#### All Submissions:

* [ ] I've signed all my commits
* [ ] I have formatted my code with [swift-format](https://github.com/apple/swift-format) per `.swift-format` [file](https://github.com/reez/BDKSwiftExampleWallet/blob/main/.swift-format)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
	- Redesigned Onboarding View with improved layout, including a new button for network selection and updated styling for buttons and text fields.
	- Introduced new views for importing wallets and configuring network settings.
	- Updated color definitions for the Accent Color to enhance visual appeal.
	- Enhanced management of Esplora server configurations with structured representations.
	- Added functionality for formatted representation of channel details.

- **Bug Fixes**
	- Removed outdated references to `ldk-node` package and streamlined package management.
	- Improved error handling during Lightning Node initialization. 

- **Chores**
	- Updated project configuration for better dependency management. 
	- Cleaned up Info.plist by removing unnecessary keys.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

reez commented:
  Thanks for the side-by-side screenshots!  

## Color

Color change looks good, but there are some additional considerations I'm thinking about it:
- current button color is due to network color, if we make the proposed change then we should concurrently make the same change across the entire app where that similar button's color was based on network color
- if we make proposed change, and don't need a network color across the onboarding/app anymore we should remove related code from view models and views, as a part of that change, if not needed anymore
- my opinion is I wouldn't mind using their dark blue (000628 from background of lightningdevkit.org in dark mode) instead of the regular blue, but I understand why you used the regular blue since it is matching the light/dark modes that are being matched here from their website. The dark blue things is strictly my opinion of what is more aesthetically pleasing to me of the two blues that is still based in the lightningdevkit website
- maybe we can have a larger discussion now or later about color scheme since we have more freedom with that and don't necessarily need to be tied to lightningdevkit website even though that is a good grounding point/basis

Overall I'm good with the proposed color change though for now, but do think it should be made across the entire app for cohesion.

## Casing

https://developer.apple.com/design/human-interface-guidelines/writing
 “Title or sentence case. Decide whether you want to use title case or sentence case for alerts, page titles, headlines, button labels, and links. Throughout the HIG, you’ll find guidelines for specific components, but how you format your text is a reflection of your app’s voice. Title case is more formal, while sentence case is more casual. Choose a style that fits your app.”

https://developer.apple.com/design/human-interface-guidelines/buttons
“Consider using text when a short label communicates more clearly than an icon. To use text, write a few words that succinctly describe what the button does. Using [title-style capitalization](https://help.apple.com/applestyleguide/#/apsgb744e4a3?sub=apdca93e113f1d64), consider starting the label with a verb to help convey the button’s action — for example, a button that lets people add items to their shopping cart might use the label “Add to Cart.”

https://www.apple.com/wallet/ 
<img width="367" alt="Screenshot 2024-10-27 at 1 06 18 PM" src="https://github.com/user-attachments/assets/4ffa73c8-af4c-4fad-a438-ce9751bd6f68">

I’m definitely not against apps that use Sentence case (one of my favorite bitcoin apps uses it), and don’t think Apple even thinks one is always better than the other, but it does seem like the strongest case for which would be the standard for an iOS wallet would be Title (per combination of HIG+Wallet)?

## UI

Does not work well on non-DynamicType or with Dynamic Type (even though I'm not as explicit about that here (yet) as I am in BDKSwiftExampleWallet), here are screenshots of proposed changes:

<img width="603" alt="Screenshot 2024-10-27 at 9 03 59 AM" src="https://github.com/user-attachments/assets/42b8f167-26d7-4700-bba2-46427a44b509">

<img width="610" alt="Screenshot 2024-10-27 at 9 04 10 AM" src="https://github.com/user-attachments/assets/902c7c37-2cce-4f97-8a67-1d29511620cd">

## Text

I'm good with changing "create wallet" from "start node", good change!

## Other

Would encourage continuing discussion on each of these as separate Issues/PR's because this PR discussion might get too large and span too many different topics (and possibly opening up Issues before PR's for these items)

Would love your help on some existing ui changes like https://github.com/reez/Monday/issues/133 or large feature experimentation/implemenation like https://github.com/reez/Monday/issues/132

I appreciate the experimentation on the `OnboardingView`!

reez reviewed (changes requested):
See prev comments above

danielnordh commented on LDKNodeMonday/View/Home/OnboardingView.swift:
```diff
@@ -90,63 +77,55 @@ struct OnboardingView: View {
+                        tintColor: .accent,
+                        isCapsule: true
+                    )
+                )
```
Added to Todo list

danielnordh commented on LDKNodeMonday/Service/Lightning Service/LightningNodeService.swift:
```diff
@@ -24,7 +24,7 @@ class LightningNodeService {
         let storedNetworkString = try! keyService.getNetwork() ?? Network.signet.description
         let storedEsploraURL =
             try! keyService.getEsploraURL()
-            ?? Constants.Config.EsploraServerURLNetwork.Signet.mutiny
+        ?? Constants.Config.EsploraServerURLNetwork.Signet.mutiny.url
 
```
Solve sepatetely

danielnordh commented on LDKNodeMonday/View/Home/OnboardingView.swift:
```diff
@@ -12,7 +12,11 @@ import SwiftUI
     @AppStorage("isOnboarding") var isOnboarding: Bool?
     @AppStorage("isFirstTime") var isFirstTime: Bool = true
-    @ObservedObject var viewModel: OnboardingViewModel
+    
```
This comment is incorrect after moving to the new (iOS 17) Observable pattern

danielnordh commented on LDKNodeMonday.xcodeproj/project.pbxproj:
```diff
@@ -173,7 +173,7 @@
 			buildActionMask = 2147483647;
 			files = (
-				AEFA825E2CC15C7C0095B0BA /* LDKNode in Frameworks */,
+				65875A932CCB9809000D3E70 /* LDKNode in Frameworks */,
```
I'm not sure why it's picking up this. Might be best if you checkout the fork and set the correct LDK-node version you want.

danielnordh commented:
Should be good for review now.
Might be worth you checking out the fork to set and commit the developer team, ldk node package and any other info that should not have changed.

reez commented on LDKNodeMonday/Utilities/Constants.swift:
```diff
@@ -14,31 +14,31 @@ struct Constants {
             struct Signet {
-                static let bdk = "http://signet.bitcoindevkit.net"
-                static let mutiny = "https://mutinynet.com/api"
+                static let bdk = EsploraServer(name: "BDK", url: "http://signet.bitcoindevkit.net")
```
we can do this as a follow up issue/pr, resolving for now

reez commented on LDKNodeMonday/Utilities/Constants.swift:
```diff
@@ -99,3 +114,8 @@ struct Constants {
 
 }
+
+struct EsploraServer: Hashable {
+    var name: String
+    var url: String
+}
```
I like it, thanks for adding-

danielnordh replied:
Improved this by moving the static definitions into this struct, which also helps fix the ?? instead of ! issues in OnboardingViewModel

reez commented on LDKNodeMonday/View Model/Home/OnboardingViewModel.swift:
```diff
@@ -8,23 +8,23 @@
                 try KeyClient.live.saveNetwork(networkString)
-                selectedURL = availableURLs.first ?? ""
-                try KeyClient.live.saveEsploraURL(selectedURL)
+                self.selectedEsploraServer = availableEsploraServers.first!
```
change requested: lets just use `??` here instead of `!`, even though maybe later we will want to handle this even better by throwing/showing an error to the user in the best way

```suggestion
                self.selectedEsploraServer = availableEsploraServers.first ?? EsploraServer(name: "", url: "")
```

reez commented on LDKNodeMonday/View Model/Home/OnboardingViewModel.swift:
```diff
@@ -8,23 +8,23 @@
-    @Published var networkColor = Color.gray
-    @Published var onboardingViewError: MondayError?
-    @Published var seedPhrase: String = "" {
+@Observable
+class OnboardingViewModel {
+    var networkColor = Color.gray
+    var onboardingViewError: MondayError?
+    var seedPhrase: String = "" {
```
change requested: I like @Observable more too, but since we AppStorage and @Observable dont play nice together I see why you did what you did. 

however I'd prefer to leave this view model using `ObservableObject` and `@AppStorage("isOnboarding") var isOnboarding: Bool?` for the moment just because we do the same thing in 2 other view models, so if we make a change to this one we should make a change to the other 2, and for the moment I'd prefer we just keep things as is to keep this PR focused (unless you just couldn't get things to work with ObservableObject at all? I know you mentioned something about ObservableObject related to this PR, so let me know if ObservableObject just didn't work for you anymore here?)

I do like using `@Observable` and I also like using `AppStorage` so we will have to come back to this in another PR to get those two working together.

danielnordh replied:
Managed to switch back to @ObservableObject, clearly the errors I encountered previously were user errors.
This minimises the changes for this PR.

I do wonder however for the future if we can find another pattern to rely on when knowing if the app should be in onboarding mode or not, other than calling Appstorage from so many different files. Some options:

- If a node is running successfully in LightningNodeService, we are not in onboarding mode
- If BackupInfo is found (for the relevant network) node should start and we are not in onboarding mode

Here's what I did in BDG Wallet:

```.swift
// Check if user already has a node seed
if backupManager.backupInfo != nil {
   // If they do, start node
   do {
      try ldkNodeManager.start(mnemonic: backupManager.backupInfo!.mnemonic, passphrase: nil)
   } catch let error {
      debugPrint(error)
   }
}
```

reez commented on LDKNodeMonday/View Model/Home/OnboardingViewModel.swift:
```diff
@@ -84,9 +86,11 @@ class OnboardingViewModel: ObservableObject {
-                self.selectedURL = esploraURL
+                self.selectedEsploraServer = availableEsploraServers.first(where: {
+                    $0.url == esploraURL
+                })!
```
change requested: lets use `??` instead of `!`

reez commented on LDKNodeMonday/View Model/Home/OnboardingViewModel.swift:
```diff
@@ -84,9 +86,11 @@ class OnboardingViewModel: ObservableObject {
+                })!
             } else {
-                self.selectedURL = availableURLs.first ?? ""
+                self.selectedEsploraServer = availableEsploraServers.first!
```
change requested: lets use `??` instead of `!`

reez commented on LDKNodeMonday/View Model/Home/OnboardingViewModel.swift:
```diff
@@ -104,10 +108,10 @@ class OnboardingViewModel: ObservableObject {
             LightningNodeService.shared = LightningNodeService()
             DispatchQueue.main.async {
-                self.isOnboarding = false
+                UserDefaults.standard.set(false, forKey: "isOnboarding")
```
note: this will change back to `self.isOnboarding = false` when you change back to ObservableObject

reez commented on LDKNodeMonday/View/Home/ImportWalletView.swift:
```diff
@@ -0,0 +1,98 @@
+struct ImportWalletView: View {
+    @Environment(\.dismiss) private var dismiss
+
+    @AppStorage("isFirstTime") var isFirstTime: Bool = true
```
requested change: could we re-use `isOnboarding` instead of adding `isFirstTime`? It would be nice to not add `isFirstTime` if we dont have to, and in my head I equate what I'm doing with `isOnboarding` (persons first time == they are onboarding) so I think you can just use `isOnboarding`? but I might be totally missing something and it not working for you if you did that.

so that would let us not have to add `isFirstTime` in 2 views:
-`ImportWalletView`
-`OnboardingView`

danielnordh replied:
Got it. This was in the original code but I never understood the difference between isOnboarding and isFirstTime.
Removed.

reez replied:
thanks, resolving conversation too

reez commented on LDKNodeMonday/View/Home/OnboardingView.swift:
```diff
@@ -23,130 +27,78 @@ struct OnboardingView: View {
-            )
-        }
+
+            }.dynamicTypeSize(...DynamicTypeSize.accessibility2)  // Sets max dynamic size for all Text
```
I like this, remind me to do this if you see me not doing it in the future on my PR's

reez commented on LDKNodeMonday/View/Home/NetworkSettingsView.swift:
```diff
@@ -0,0 +1,78 @@
+            .navigationTitle("Network settings")
+            .navigationBarTitleDisplayMode(.inline)
+            .scrollContentBackground(.hidden)
+            .toolbar {
```
change requested: remove this entire `.toolbar` because the `Back` button does not make sense to me, since when we dismiss this sheet it dismisses down (not left). 

I think we chatted about this a bit, and thought maybe a "Done" in the toolbar trailing might be the thing, so feel free to do that if you'd like but the "Back" button is something I'd definitely not keep. I'm ok with it just having a swipe to dismiss right now and we come back to the "Done" part later if we're not 100% sure what we want to do with that part at the moment.

danielnordh replied:
Instead of removing and having no visual dismiss button here I have replaced the Back button on the left with a Done button on the right, following iOS conventions.

reez replied:
perfect, resolving conversation

reez reviewed (changes requested):
This is looking good, some comments about unimportant things, some comments about nice things, some `changes requested:` comments for changes I’m requesting.

Some of the changes requested are not even that what you have is wrong and what I have is far superior, but might just be that it breaks convention with the current codebase, and I don’t want you having to change it in other unrelated places in the code base if we are changing the convention right now (would rather just do that in follow up PR’s to allow us to not expand the size of the changes in this PR so this PR can be merged sooner rather than later).

danielnordh commented:
Thanks for the review, good points.
Given the history is getting long and unwieldy with all this CodeRabbit comments, here is what I've changed:

- Went back to @ObservableObject for OnboardingViewModel
- Replaced the "< Back" Toolbar pattern on the left with Done/Cancel on the right
- Replaced ! with ?? as requested
- Removed unnecessary isFirstTime variable

danielnordh commented on LDKNodeMonday.xcodeproj/project.pbxproj:
```diff
@@ -86,11 +90,13 @@
 /* Begin PBXFileReference section */
+		658D1E822CFF1978005BD1F3 /* NetworkSettingsView.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = NetworkSettingsView.swift; sourceTree = "<group>"; };
+		659EE6942CF8D4990064ED78 /* ImportWalletView.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = ImportWalletView.swift; sourceTree = "<group>"; };
+		65A406622D070CBD00EA331E /* NetworkSettingsView.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = NetworkSettingsView.swift; sourceTree = "<group>"; };
```
Fixed

danielnordh commented on LDKNodeMonday.xcodeproj/project.pbxproj:
```diff
@@ -296,9 +302,12 @@
 				AE186B8D2A1540B700338463 /* StartView.swift */,
 				AE4C15B12A1316D100AA8917 /* BitcoinView.swift */,
+				659EE6942CF8D4990064ED78 /* ImportWalletView.swift */,
+				658D1E822CFF1978005BD1F3 /* NetworkSettingsView.swift */,
```
Fixed

danielnordh commented on LDKNodeMonday.xcodeproj/project.pbxproj:
```diff
@@ -7,6 +7,10 @@
 
 /* Begin PBXBuildFile section */
+		65875A932CCB9809000D3E70 /* LDKNode in Frameworks */ = {isa = PBXBuildFile; productRef = 65875A922CCB9809000D3E70 /* LDKNode */; };
+		658D1E832CFF1978005BD1F3 /* NetworkSettingsView.swift in Sources */ = {isa = PBXBuildFile; fileRef = 658D1E822CFF1978005BD1F3 /* NetworkSettingsView.swift */; };
```
Fixed

reez commented:
Yes this looks good! only thing left for approval is the merge conflict, once that is resolved I will approve and merge-
